test/app-luatest/key_estimator_test.lua

```diff
@@ -0,0 +1,110 @@
+local t = require('luatest')
+local g = t.group()
+local key_def = require('key_def')
+local key_estimator = require('key_estimator')
+
+local function hll_error(prec)
+    return 1.04 / (2 ^ (prec / 2.0))
+end
+
+g.test_constants = function()
+    t.assert_not_equals(key_estimator.SPARSE, key_estimator.DENSE)
+    t.assert_lt(key_estimator.MIN_PRECISION, key_estimator.MAX_PRECISION)
+end
+
+g.test_common = function()
+    --- expect key format
+    t.assert_error(key_estimator.new)
+
+    local fmt = key_def.new({{type = "string", fieldno = 1}})
+    local est = key_estimator.new(fmt)
+
+    --- invalid precision
+    t.assert_error(key_estimator.new, fmt, key_estimator.MAX_PRECISION + 1)
+
+    t.assert_equals(est:estimate(), 0)
+    est:add({'bla'})
+    t.assert_equals(est:estimate(), 1)
+    est:add({'bla'})
+    t.assert_equals(est:estimate(), 1)
+    est:add({'bla', 'bla', 'bla..'})
+    t.assert_equals(est:estimate(), 1)
+
+    --- wrong key format, string is expected
+    t.assert_error(key_estimator.add, est, {0})
+end
+
+g.test_error = function()
+    local fmt = key_def.new({{type = "string", fieldno = 1}})
+
+    local min_prec = key_estimator.MIN_PRECISION
+    local max_prec = key_estimator.MAX_PRECISION
+    for prec = min_prec, max_prec, 1 do
+        local est = key_estimator.new(fmt, prec)
+        local err = hll_error(prec)
+
+        local card = 10 * 2 ^ prec
+        for i = 0, card, 1 do
+            est:add({tostring(i)})
+        end
+
+        --- 99.7% to lie in expected interval
+        local margin = card * err * 3
+        t.assert_almost_equals(est:estimate(), card, margin)
+    end
+end
+
+g.test_merge = function()
+    local fmt = key_def.new({{type = "string", fieldno = 1}})
+
+    local min_prec = key_estimator.MIN_PRECISION
+    local max_prec = key_estimator.MAX_PRECISION
+    for prec = min_prec, max_prec, 1 do
+        local est1 = key_estimator.new(fmt, prec, key_estimator.DENSE)
+        local est2 = key_estimator.new(fmt, prec, key_estimator.DENSE)
+        local err = hll_error(prec)
+
+        local card = 5 * 2 ^ prec
+        for i = 0, card, 1 do
+            if i % 2 == 0 then
+                est1:add({tostring(i)})
+            elseif i % 3 == 0 then
+                est2:add({tostring(i)})
+            else
+                est1:add({tostring(i)})
+                est2:add({tostring(i)})
+            end
```

Comment:
I don't quite understand this logic.

Reply:
For this test, I divided the set range [0, map] between 2 estimators. Some items belong to only one estimator, and some belong to both of them. For the completeness of the test, we need to check that adding common elements does not change the estimation, unlike adding new ones. This is achieved by merging estimators that have some  common items and some unique items.